src/mono/mono/metadata/native-library.c

```diff
@@ -1610,14 +1610,26 @@ ves_icall_System_Runtime_InteropServices_NativeLibrary_GetSymbol (gpointer lib,
 	native_library_lock ();
 
 	module = netcore_handle_lookup (lib);
-	if (!module)
+	if (!module) {
+#ifdef HOST_WIN32
+		// netcore calls NativeLibrary.GetExport on Windows system libraries loaded directly using Interop.Kernel32.LoadLibraryEx.
+		// Fallback accepting NativeLibrary.GetExport calls on libraries loaded directly by Interop.Kernel32.LoadLibraryEx.
```

Comment:
So the issue is: Mono needs to associate extra info with the native module handle (in the case that an embedder is using hooks to modify native library lookups), but `LoadLibraryEx` will allow the managed code (CoreLib, in this case) to sidestep the embedder. (@CoffeeFlux this doesn't seem windows specific - if CoreLib had `Interop.Unix.Dlopen` it would be the same problem, right?)

On the other hand, we don't guarantee to the embedder that their hook is the only one installed, or that they will be called for everything (ie an earlier hook could resolve the module, and a later one won't be called - right?).

So we could (either logically, or actually) associate handles returned by `LoadLibraryEx` with a Windows-specific hook that happens to run before the embedder's hook:

- One way we could do that is by changing `Interop.Kernel32.LoadLibraryEx` from being a PInvoke to being a small wrapper around a PInvoke (or a qcall into mono to forward to LoadLibraryEx and also set up the association).

- another option is to treat an un-associated `lib` in `NativeLibrary.GetSymbol` as though it came from the LoadLibraryEx hook.  (which btw is my one suggestion for this PR: in the `!module` case here, make subsequent `netcore_handle_lookup(lib)` calls return some pre-fabricated `MonoDl` instance that just calls `GetProcAddress`.

Reply:
Yes, that is what I'm leaning towards as well, return something that is a dummy MonoDl that will just make sure we can get the library handle passed to the implemented OS specific API's in mono_dl_symbol or mono_dl_close, writing something up at the moment.

Reply:
This is indeed not Windows specific.

Creating a `MonoDl` in the failure case seems like a reasonable option to me. 

One thing to note is that we currently handle the ref counting at the `MonoDl` level rather than relying on the OS APIs, so intermingling direct syscalls with `NativeLibrary` on the same library is always going to be a bug farm when it comes to opening/freeing. I am still inclined to say we should discourage that, but this will at least enable basic usage outside of freeing to work correctly.

Reply:
I also see that we have a block list when we unload, so maybe the most safe way is to not mask it into the logic of that cache, but keep it in:

ves_icall_System_Runtime_InteropServices_NativeLibrary_GetSymbol
ves_icall_System_Runtime_InteropServices_NativeLibrary_FreeLib

when we fail to load library from cache, and use a dummy MonoDl just carrying the OS library handle as needed by direct calls to mono_dl_symbol/mono_dl_close.

Will make a proposal of that in this PR.

Reply:
Also note that there is also the other direction: Somebody may get the handle via `NativeLibrary` and pass it to the OS loader directly, e.g. to get versioned symbol via `dlvsym`.

Reply:
Yeah, and that should work - I think we even test that particular case. We return the OS handle directly, so passing it to other libraries to call into works fine.